patches/api/0387-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,1070 @@
++        builder.setProjectile(true).setFire(true);
++        return builder.build();
++    }
++
++    @NotNull
++    static IndirectEntityDamageOrigin ofWitherSkull(@NotNull WitherSkull projectile) {
++        return ofWitherSkull(projectile, null);
++    }
++
++    @NotNull
++    static IndirectEntityDamageOrigin ofWitherSkull(@NotNull WitherSkull projectile, @Nullable Entity damager) {
++        return indirectEntitySource("witherSkull", projectile, damager).setProjectile(true).build();
++    }
++
++    @NotNull
++    static IndirectEntityDamageOrigin ofPassiveProjectile(@NotNull Entity projectile) {
++        return ofPassiveProjectile(projectile, null);
++    }
++
++    @NotNull
++    static IndirectEntityDamageOrigin ofPassiveProjectile(@NotNull Entity projectile, @Nullable Entity damager) {
++        return indirectEntitySource("thrown", projectile, damager).setProjectile(true).build();
++    }
++
++    @NotNull
++    static IndirectEntityDamageOrigin ofIndirectMagic(@NotNull Entity entity) {
++        return ofIndirectMagic(entity, null);
++    }
++
++    @NotNull
++    static IndirectEntityDamageOrigin ofIndirectMagic(@NotNull Entity entity, @Nullable Entity damager) {
++        return indirectEntitySource("indirectMagic", entity, damager).setMagic(true).bypass(DamageBypass.ARMOR).build();
++    }
++
++    @NotNull
++    static EntityDamageOrigin ofThorns(@NotNull Entity damager) {
++        return entitySource("thorns", damager).setMagic(true).setThorns(true).build();
++    }
++
++    @NotNull
++    static DamageOrigin ofExplosion() {
++        return ofExplosion(null);
++    }
++
++    @NotNull
++    static DamageOrigin ofExplosion(@Nullable LivingEntity source) {
++        if(source != null) {
++            return entitySource("explosion.player", source).setExplosion(true).scalesWithDifficulty(true).build();
++        }
++        return of("explosion").setExplosion(true).scalesWithDifficulty(true).build();
++    }
++
++    @NotNull
++    static DamageOrigin ofSonicBoom(@NotNull Entity damager) {
++        return entitySource("sonic_boom", damager).setMagic(true).bypassOf(DamageBypass.ARMOR, DamageBypass.ENCHANTMENT).build();
++    }
++
++    boolean isSweep();
++
++    boolean isCritical();
++
++    boolean isFromProjectile();
++
++    boolean isFromExplosion();
++
++    boolean isFromFire();
++
++    boolean isFromMagic();
++
++    boolean isFromFalling();
++
++    boolean willNotAggro();
++
++    boolean willScalesWithDifficulty();
++
++    default boolean canBypass(@NotNull DamageBypass... damageType) {
++        for(DamageBypass type : damageType) {
++            if(!canBypass(type)) return false;
++        }
++        return true;
++    }
++
++    boolean canBypass(@NotNull DamageBypass damageType);
++
++    boolean applyArmorExhaustion();
++
++    boolean applyMagicExhaustion();
++
++    boolean canDamageHelmetArmor();
```

Comment:
See my whole spiel about the bypass enum.
Also, there shouldn't be mutable methods in here.